File: src/ZKTecoADMS.Domain/Entities/PosCustomer.cs
using System.ComponentModel.DataAnnotations;
using ZKTecoADMS.Domain.Entities.Base;

namespace ZKTecoADMS.Domain.Entities;

/// <summary>Khách hàng POS (CRM).</summary>
public class PosCustomer : AuditableEntity<Guid>
{
    [Required]
    public Guid StoreId { get; set; }
    public virtual Store? Store { get; set; }

    [Required]
    [MaxLength(30)]
    public string CustomerCode { get; set; } = string.Empty;

    [Required]
    [MaxLength(200)]
    public string Name { get; set; } = string.Empty;

    [MaxLength(50)]
    public string? Phone { get; set; }

    [MaxLength(200)]
    public string? Email { get; set; }

    [MaxLength(500)]
    public string? Address { get; set; }

    [MaxLength(100)]
    public string? Province { get; set; }

    [MaxLength(100)]
    public string? Ward { get; set; }

    [MaxLength(200)]
    public string? CompanyName { get; set; }

    [MaxLength(50)]
    public string? TaxCode { get; set; }

    public DateTime? Birthday { get; set; }

    [MaxLength(500)]
    public string? DeliveryAddress { get; set; }

    [MaxLength(1000)]
    public string? Note { get; set; }

    /// <summary>Tổng giá trị đã mua (đơn hoàn thành).</summary>
    public decimal TotalPurchase { get; set; }

    /// <summary>Công nợ khách hàng hiện tại.</summary>
    public decimal CurrentDebt { get; set; }

    /// <summary>Điểm tích lũy hiện tại.</summary>
    public decimal PointBalance { get; set; }

    public virtual ICollection<PosSaleOrder> SaleOrders { get; set; } = [];
    public virtual ICollection<PosCustomerPayment> Payments { get; set; } = [];
    public virtual ICollection<PosCustomerPointTransaction> PointTransactions { get; set; } = [];
}

File: flutter_client/lib/models/pos_customer.dart
class PosCustomer {
  final String id;
  final String customerCode;
  final String name;
  final String? phone;
  final String? email;
  final String? address;
  final String? province;
  final String? ward;
  final String? companyName;
  final String? taxCode;
  final String? note;
  final double totalPurchase;
  final double currentDebt;

  PosCustomer({
    required this.id,
    required this.customerCode,
    required this.name,
    this.phone,
    this.email,
    this.address,
    this.province,
    this.ward,
    this.companyName,
    this.taxCode,
    this.note,
    this.totalPurchase = 0,
    this.currentDebt = 0,
  });

  factory PosCustomer.fromJson(Map<String, dynamic> json) {
    double n(dynamic v) => v is num ? v.toDouble() : double.tryParse('$v') ?? 0;
    return PosCustomer(
      id: (json['id'] ?? json['Id'] ?? '').toString(),
      customerCode: (json['customerCode'] ?? json['CustomerCode'] ?? '').toString(),
      name: (json['name'] ?? json['Name'] ?? '').toString(),
      phone: json['phone'] ?? json['Phone'] as String?,
      email: json['email'] ?? json['Email'] as String?,
      address: json['address'] ?? json['Address'] as String?,
      province: json['province'] ?? json['Province'] as String?,
      ward: json['ward'] ?? json['Ward'] as String?,
      companyName: json['companyName'] ?? json['CompanyName'] as String?,
      taxCode: json['taxCode'] ?? json['TaxCode'] as String?,
      note: json['note'] ?? json['Note'] as String?,
      totalPurchase: n(json['totalPurchase'] ?? json['TotalPurchase']),
      currentDebt: n(json['currentDebt'] ?? json['CurrentDebt']),
    );
  }
}
